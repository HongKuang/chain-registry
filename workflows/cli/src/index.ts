export const cli = 1;