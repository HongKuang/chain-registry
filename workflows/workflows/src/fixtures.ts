import { sync as glob } from "glob";
import { basename, dirname, join, relative } from "path";

export const NON_INFO_DIRS: string[] = [
  '_memo_keys',
  '_scripts',
  '_template',
  '.github'
];
const REG_DIR: string = 'chain-registry/chain-registry';
export const registryDir: string = join(__dirname, `/../../../packages/${REG_DIR}`);

// Type definitions for file path processing results
export interface FilePathInfo {
  filepath: string;
  origpath: string;
}

// Helper function to filter and map file paths
function filterAndMapFiles(paths: string[]): FilePathInfo[] {
  return paths.filter(path => {
    const relativePath: string = path.split(REG_DIR)[1];
    const dir: string = basename(dirname(relativePath));
    return !NON_INFO_DIRS.includes(dir);
  }).map(path => {
    const relativePath: string = path.split(REG_DIR)[1];
    return {
      filepath: relativePath,
      origpath: relative(process.cwd(), path)
    };
  });
}

// Glob patterns for different JSON files
const globPatterns: { [key: string]: string } = {
  chains: `${registryDir}/**/chain.json`,
  assetLists: `${registryDir}/**/assetlist.json`,
  ibcInfo: `${registryDir}/**/_IBC/*-*.json`,
  schemas: `${registryDir}/*.schema.json`
};

// Using the helper function to process files
export const chains: FilePathInfo[] = filterAndMapFiles(glob(globPatterns.chains));
export const assetLists: FilePathInfo[] = filterAndMapFiles(glob(globPatterns.assetLists));
export const ibcInfo: FilePathInfo[] = filterAndMapFiles(glob(globPatterns.ibcInfo));
export const schemas: FilePathInfo[] = filterAndMapFiles(glob(globPatterns.schemas));
