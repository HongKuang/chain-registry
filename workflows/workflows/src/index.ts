export * from './registry';
export * from './schema-typescript';
export { JSONSchema } from 'schema-typescript';
