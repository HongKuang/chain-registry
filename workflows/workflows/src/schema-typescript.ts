import * as fs from 'fs';
import { basename, dirname, join } from 'path';
import { generateTypeScript, JSONSchema, SchemaTSOptions } from 'schema-typescript';

import { FilePathInfo } from './fixtures';

// Default titles for certain schemas
// TODO create issue in cosmos/chain-registry
const DEFAULT_TITLES: { [filename: string]: string } = {
  'ibc_data.schema.json': 'IBCData',
  'chain.schema.json': 'Chain',
  'assetlist.schema.json': 'AssetList',
  'versions.schema.json': 'Versions',
  'memo_keys.schema.json': 'MemoKeys',
};


// Define an interface for the constructor options
export interface SchemaTypeGeneratorOptions {
  outputDir: string;
  schemaTSOptions: Partial<SchemaTSOptions>;
  schemas: FilePathInfo[];
  supportedSchemas?: string[];
}

export class SchemaTypeGenerator {
  private outputDir: string;
  private schemas: FilePathInfo[];
  private schemaTSOptions: Partial<SchemaTSOptions>
  private supportedSchemas: Set<string>;

  constructor(options: SchemaTypeGeneratorOptions) {
    this.outputDir = options.outputDir;
    this.schemas = options.schemas;
    this.supportedSchemas = new Set(options.supportedSchemas || []);
  }

  private updateSchemaTitle(schema: JSONSchema, schemaFile: string): void {
    const filename = basename(schemaFile);
    if (DEFAULT_TITLES[filename]) {
      schema.title = DEFAULT_TITLES[filename];
    }
  }

  private isSchemaSupported(filePath: string): boolean {
    if (this.supportedSchemas.size === 0) return true; // If no filter is provided, process all files
    const filename = basename(filePath);
    return this.supportedSchemas.has(filename);
  }

  public generateTypes(): void {
    this.schemas.forEach(fileInfo => {
      const schemaFile = fileInfo.origpath;
      if (this.isSchemaSupported(schemaFile)) {
        try {
          const schema: JSONSchema = this.readJsonFile(schemaFile);
          this.updateSchemaTitle(schema, schemaFile);
          const result = this.generateTypeScript(schema);
          const filename = this.getOutputFilename(schemaFile);
          this.ensureDirExists(filename);
          this.writeFile(filename, result);
        } catch (e) {
          console.log({ schemaFile })
        }
      }
    });
  }

  private readJsonFile(filePath: string): JSONSchema {
    return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
  }

  private generateTypeScript(schema: JSONSchema): string {
    return generateTypeScript(schema, this.schemaTSOptions);
  }

  private getOutputFilename(schemaFile: string): string {
    const filename = basename(schemaFile);
    return join(this.outputDir, filename.replace(/.json$/, '.ts'));
  }

  private ensureDirExists(filePath: string): void {    
    this.mkdirpSync(dirname(filePath));
  }

  private mkdirpSync(p: string): void {
    if (!fs.existsSync(p)) {
      this.mkdirpSync(dirname(p));
      fs.mkdirSync(p);
    }
  }


  private writeFile(filePath: string, content: string): void {
    fs.writeFileSync(filePath, content);
  }
}
