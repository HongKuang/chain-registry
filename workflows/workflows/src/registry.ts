import { AssetList, Chain, IBCData, MemoKeys, Versions } from '@chain-registry/interfaces';
import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { basename, join } from 'path';
import { JSONSchema } from "schema-typescript";

const SCHEMATA_MAPPING: Record<string, string> = {
  AssetList: 'assetLists',
  Chain: 'chains',
  IBCData: 'ibcData',
  MemoKeys: 'memoKeys',
  Versions: 'versions'
};

const SCHEMA_WHITELIST = [
  'assetlist.schema.json',
  'chain.schema.json',
  'ibc_data.schema.json',
  'memo_keys.schema.json',
  'versions.schema.json'
];

interface JSONSchemaContent {
  $schemaFile: string;
  path: string;
  content: JSONSchema;
}

interface DataMapping {
  [key: string]: any[];
}

export class RegistryFixture {
  private definitions: JSONSchemaContent[];
  private basePath: string;
  private schemaMappings: { [title: string]: JSONSchema } = {};
  private dataMappings: DataMapping = {};

  constructor(basePath: string) {
    this.basePath = basePath;
    this.initializeData();
  }

  private getCamelCase(schema: JSONSchema): string {
    return SCHEMATA_MAPPING[schema.title];
  }

  private isJsonSchema(schema: JSONSchema): boolean {
    const $schema = schema.$schema;
    if (!$schema) return false;
    return true;
  }

  private startsWithUrl(schema: JSONSchema): boolean {
    return (/^http/.test(schema.$schema))
  }

  private isSchema(schema: JSONSchema): boolean {
    if (!this.isJsonSchema(schema)) return false;
    if (!this.startsWithUrl(schema)) return false;
    const url = new URL(schema.$schema);
    const def = url.pathname;
    if (def.match(/\/draft-0[47]\/schema$/)) {
      return true;
    }
    return false;
  }

  private getSchemaFrom$Schema(data: JSONSchemaContent): JSONSchema {
    const schema = this.definitions.find(defn => {
      const pred = basename(defn.path) === data.$schemaFile;
      return pred;
    });
    if (!schema) {
      console.warn('missing schema definition: ' + data.$schemaFile);
    }
    return schema.content;
  }

  private isData(schema: JSONSchema): boolean {
    if (!this.isJsonSchema(schema)) return false;
    if (this.startsWithUrl(schema)) return false;
    if (SCHEMA_WHITELIST.includes(basename(schema.$schema))) return true;
    return false;
  }

  private loadData(schema: JSONSchemaContent): void {
    if (this.isData(schema.content)) {
      const defn = this.getSchemaFrom$Schema(schema);
      this.dataMappings[this.getCamelCase(defn)].push(schema.content);
    }
  }

  private initStoreForSchema(schema: JSONSchema) {
    // if it exists return
    if (schema.title in this.schemaMappings) return;

    // initialize
    const camel = this.getCamelCase(schema);
    this.schemaMappings[schema.title] = schema;
    this.dataMappings[camel] = [];
  }

  private initializeData(): void {

    let types: any = {};

    // parse every JSON file
    this.definitions = glob(`${this.basePath}/**/*.json`)
      .map(path => {
        const content = JSON.parse(readFileSync(path, 'utf-8'));
        if (!this.isJsonSchema(content)) return;
        types[basename(content.$schema)] = true;
        return {
          $schemaFile: basename(content.$schema),
          path,
          content
        };
      }).filter(Boolean)

    // filter out schemas (e.g. draft-04/schema )
    const schemas = this.definitions.filter(schema => this.isSchema(schema.content)).filter(Boolean);

    // validate
    schemas.forEach(schema => {
      if (!(schema.content.title in SCHEMATA_MAPPING)) throw new Error('MISSING SCHEMA: ' + schema.content.title)
    });

    // create schemaMappings and data
    schemas.forEach(schema => {
      this.initStoreForSchema(schema.content);
    });

    // load data
    this.definitions
      .filter(d => SCHEMA_WHITELIST.includes(d.$schemaFile))
      .forEach(schema => {
        this.loadData(schema)
      })
  }

  public get chains(): Chain[] {
    return this.dataMappings.chains || [];
  }

  public get assetLists(): AssetList[] {
    return this.dataMappings.assetLists || [];
  }

  public get ibcData(): IBCData[] {
    return this.dataMappings.ibcData || [];
  }

  public get memoKeys(): MemoKeys[] {
    return this.dataMappings.memoKeys || [];
  }

  public get versions(): Versions[] {
    return this.dataMappings.versions || [];
  }

  public get count() {
    return {
      chains: this.chains.length,
      assetLists: this.assetLists.length,
      ibcData: this.ibcData.length,
      memoKeys: this.memoKeys.length,
      versions: this.versions.length,
    }
  }
}