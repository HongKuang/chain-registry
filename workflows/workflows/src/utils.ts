import { toCamelCase, toPascalCase } from 'schema-typescript';

export const camelCase = (str: string) => {
  if (str === 'IBCData') {
    return 'ibcData';
  }
  return toCamelCase(str);
};

export const pascalCase = (str: string) => {
  return toPascalCase(str);
};