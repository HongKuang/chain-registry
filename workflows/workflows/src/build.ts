import { writeFileSync } from 'fs';
import { sync as mkdirp } from 'mkdirp';
import { dirname, join } from 'path';
import { jsonStringify, JSONStringifyOptions } from 'strfy-js';

import { JSONSchemaContent, Registry } from './registry';

import { JSONSchemaPatch, JSONSchemaPatchOperation } from 'json-schema-patch';

export interface RegistryBuilderOptions {
  assetList: JSONStringifyOptions;
  chain: JSONStringifyOptions;
  ibcData: JSONStringifyOptions;
  
  ops: {
    assetList: JSONSchemaPatchOperation[];
    chain: JSONSchemaPatchOperation[];
    ibcData: JSONSchemaPatchOperation[];
  }
  
}

export class RegistryBuilder {
  options: RegistryBuilderOptions;
  private registry: Registry;

  constructor(registry: Registry, options?: RegistryBuilderOptions) {
    this.registry = registry;
    this.options = options;
  }

  writeFile(outDir: string, info: JSONSchemaContent<any>, options: JSONStringifyOptions) {
    const str = jsonStringify(info.content, options);
    const file = join(outDir, info.path.replace(this.registry.basePath, ''));
    mkdirp(dirname(file));
    writeFileSync(file, str);
  }

  writeSchema(outDir: string, info: JSONSchemaContent<any>, ops: JSONSchemaPatchOperation[]) {
    const patcher = new JSONSchemaPatch(info.content);
    ops.forEach(patcher.prepareOperation.bind(patcher));
    patcher.applyPatch();

    const file = join(outDir, info.path.replace(this.registry.basePath, ''));
    mkdirp(dirname(file));
    writeFileSync(file, JSON.stringify(patcher.schema, null, 2));
  }

  buildSchemas(outDir: string) {
    this.writeSchema(outDir, this.registry.schemaMappings.AssetList, this.options.ops.assetList);
    this.writeSchema(outDir, this.registry.schemaMappings.Chain, this.options.ops.chain);
    this.writeSchema(outDir, this.registry.schemaMappings.IBCData, this.options.ops.ibcData);
  }

  build(outDir: string) {
    this.registry.dataMappings.AssetList.forEach(info => {
      this.writeFile(outDir, info, this.options.assetList);
    });
    this.registry.dataMappings.Chain.forEach(info => {
      this.writeFile(outDir, info, this.options.chain);
    });
    this.registry.dataMappings.IBCData.forEach(info => {
      this.writeFile(outDir, info, this.options.ibcData);
    });
  }

}
