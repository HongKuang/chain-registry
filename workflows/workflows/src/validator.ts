import AjvDraft07, { ValidateFunction as ValidateFunctionDraft07 } from 'ajv';
import Ajv2019, { ValidateFunction as ValidateFunction2019 } from 'ajv/dist/2019';
import Ajv2020, { ValidateFunction as ValidateFunction2020 } from 'ajv/dist/2020';
import addFormats from 'ajv-formats';
import chalk from 'chalk';
import { JSONSchema } from 'schema-typescript';

import { DataMapping, JSONSchemaContent, Registry } from './registry';

export interface SchemaValidatorOptions {
    draft?: '2019-09' | '2020-12' | 'draft-07'
    useStrict?: boolean;
    allErrors?: boolean;
    useDefaults?: boolean;
}

export class SchemaValidator {
    private ajv: Ajv2020 | Ajv2019;
    private registry: Registry;
    private options: SchemaValidatorOptions;

    constructor(registry: Registry, options?: SchemaValidatorOptions) {
        const { useStrict = false, allErrors = true, useDefaults = true, draft = '2019-09' } = options ?? {};

        this.options = {
            useDefaults,
            useStrict,
            draft,
            allErrors
        };

        switch (draft) {
            case 'draft-07':
                this.ajv = new AjvDraft07({
                    strict: useStrict,
                    allErrors,
                    useDefaults
                });
                break;
            case '2019-09':
                this.ajv = new Ajv2019({
                    strict: useStrict,
                    allErrors,
                    useDefaults
                });
                break;
            case '2020-12':
                this.ajv = new Ajv2020({
                    strict: useStrict,
                    allErrors,
                    useDefaults
                });
                break;
            default:
                throw new Error('JSONSchema draft not yet supported.')

        }


        this.registry = registry;
        addFormats(this.ajv);
    }

    public validateAllData(verbose: boolean = false) {
        // Compile and validate each schema, then validate corresponding data
        this.registry.forEachSchemas(([title, schema]) => {
            try {
                
                switch (this.options.draft) {
                    case 'draft-07':
                        schema.content.$schema = 'http://json-schema.org/draft-07/schema';
                        break;
                    case '2019-09':
                        schema.content.$schema = 'https://json-schema.org/draft/2019-09/schema';
                        break;
                    case '2020-12':
                        schema.content.$schema = 'https://json-schema.org/draft/2020-12/schema';
                        break;
                    default:
                        break;
                }
                
                const validate: any = this.ajv.compile(schema.content);

                const dataMap = this.registry.dataMappings[title as keyof DataMapping];

                if (!dataMap) {
                    console.error(chalk.yellow(`⚠️  No data found for schema titled ${chalk.bold(title)}`));
                    return;
                }

                dataMap.forEach(data => {
                    this.validateJsonSchema(data, title, validate, verbose);
                });
            } catch (e) {
                console.error(chalk.red(`❌ Strict Validation errors for schema ${chalk.bold(title)} in file ${chalk.magenta(schema.path)}:`));
                throw e;
            }
        });
    }

    private validateJsonSchema(
        data: JSONSchemaContent<JSONSchema>,
        title: string,
        validate: ValidateFunction2019<unknown> | ValidateFunction2020<unknown> | ValidateFunctionDraft07<unknown>,
        verbose: boolean
    ) {
        if (!validate(data.content)) {
            console.error(chalk.red(`❌ Validation errors for ${chalk.bold(title)} in file ${chalk.magenta(data.path)}:`));
            validate.errors?.forEach(error => {
                console.error(chalk.red(`  ➡️ ${error.instancePath} ${error.message}`));
            });
        } else if (verbose) {
            console.log(chalk.green(`✅ Validation passed for ${chalk.bold(title)} in file ${chalk.magenta(data.path)}`));
        }
    }
}
