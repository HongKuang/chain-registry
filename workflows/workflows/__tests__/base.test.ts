
import { sync as globSync } from 'glob';
import { resolve } from 'path';

import { schemas } from '../src';
import { SchemaTypeGenerator } from '../src';

const outputDir = resolve(__dirname +'/../../../__output__/base-types');

it('types', () => {
  const generator = new SchemaTypeGenerator({
    outputDir,
    supportedSchemas: [
        'chain.schema.json',
        'assetlist.schema.json',
        'ibc_data.schema.json',
        'versions.schema.json',
        'memo_keys.schema.json',
    ],
    schemas,
    schemaTSOptions: {
        strictTypeSafety: true,
        useCamelCase: false,
        useSingleQuotes: true
    }
  });
  generator.generateTypes();

  const tsFiles = globSync(`${outputDir}/*.ts`).map(a=>a.split(outputDir)[1])
  expect(tsFiles).toMatchSnapshot();
  
});