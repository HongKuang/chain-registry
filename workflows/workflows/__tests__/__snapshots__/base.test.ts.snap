// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`types 1`] = `
[
  "/assetlist.schema.ts",
  "/chain.schema.ts",
  "/ibc_data.schema.ts",
  "/memo_keys.schema.ts",
  "/versions.schema.ts",
]
`;
